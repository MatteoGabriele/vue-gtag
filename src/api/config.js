import query from "@/api/query";
import { getOptions } from "@/options";

export default (...args) => {
  const { config, includes } = getOptions();

  query("config", config.id, ...args);

  if (Array.isArray(includes)) {
    includes.forEach((domain) => {
      query("config", domain.id, ...args);
    });
  }
};
