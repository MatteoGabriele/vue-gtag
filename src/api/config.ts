import query from "@/api/query";
import { getSettings } from "@/settings";
import type { Tail } from "@/types";

type ConfigParams = Tail<Gtag.GtagCommands["config"]>;

export default function config(...args: ConfigParams) {
  const { tagId, additionalAccounts } = getSettings();

  if (!tagId) {
    return;
  }

  query("config", tagId, ...args);

  if (!additionalAccounts) {
    return;
  }

  for (const account of additionalAccounts) {
    query("config", account.tagId, ...args);
  }
}
