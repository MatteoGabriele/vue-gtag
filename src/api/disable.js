import { getOptions } from "@/options";

const assignGlobalProperty = (id, value) => {
  window[`ga-disable-${id}`] = value;
};

export default (value = true) => {
  const { config, includes } = getOptions();

  assignGlobalProperty(config.id, value);

  if (Array.isArray(includes)) {
    includes.forEach((domain) => assignGlobalProperty(domain.id, value));
  }
};
