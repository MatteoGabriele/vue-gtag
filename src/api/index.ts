export { default as config } from "@/api/config";
export { default as event } from "@/api/event";
export { default as linker } from "@/api/linker";
export { default as pageview } from "@/api/pageview";
export { default as purchase } from "@/api/purchase";
export { default as query } from "@/api/query";
export { default as refund } from "@/api/refund";
export { default as screenview } from "@/api/screenview";
export { default as time } from "@/api/time";
