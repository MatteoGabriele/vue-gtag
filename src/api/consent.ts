import { query } from "@/api/query";
import type { GtagConsentArg, GtagConsentParams } from "@/types/gtag";

export function consent(consentArg: GtagConsentArg, params: GtagConsentParams) {
  query("consent", consentArg, params);
}

export function consentGrantedAll(mode: GtagConsentArg = "default") {
  consent(mode, {
    ad_user_data: "granted",
    ad_personalization: "granted",
    ad_storage: "granted",
    analytics_storage: "granted",
  });
}

export function consentDeniedAll(mode: GtagConsentArg = "default") {
  consent(mode, {
    ad_user_data: "denied",
    ad_personalization: "denied",
    ad_storage: "denied",
    analytics_storage: "denied",
  });
}
