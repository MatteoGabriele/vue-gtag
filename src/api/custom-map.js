import config from "@/api/config";

export default (map) => {
  config({
    custom_map: map,
  });
};
