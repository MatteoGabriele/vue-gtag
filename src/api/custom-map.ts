import type { GtagCustomParams } from "src/types/gtag";
import config from "./config";

export default function customMap(params: GtagCustomParams) {
  config({
    custom_map: params,
  });
}
