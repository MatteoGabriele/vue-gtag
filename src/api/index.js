import query from "./query";
import config from "./config";
import event from "./event";
import pageview from "./pageview";
import screenview from "./screenview";
import customMap from "./custom-map";
import time from "./time";
import exception from "./exception";
import linker from "./linker";
import purchase from "./purchase";
import refund from "./refund";
import set from "./set";
import optIn from "./opt-in";
import optOut from "./opt-out";

export default {
  query,
  config,
  event,
  pageview,
  screenview,
  customMap,
  time,
  exception,
  linker,
  purchase,
  refund,
  set,
  optIn,
  optOut
};
