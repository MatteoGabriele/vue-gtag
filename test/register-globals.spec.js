import { createLocalVue } from "@vue/test-utils";
import VueGtag from "@/index";
import flushPromises from "flush-promises";
import * as api from "@/api";
import mockDate from "mockdate";

mockDate.set("2021-04-20 10:00:00");

jest.mock("@/api");

describe("registerGlobals", () => {
  beforeEach(() => {
    delete window.gtag;
    delete window.dataLayer;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("register global dataLayer and globalObject", () => {
    const localVue = createLocalVue();

    localVue.use(VueGtag, {
      config: {
        id: 1,
      },
    });

    expect(window.gtag).toBeDefined();
    expect(window.dataLayer).toBeDefined();
  });

  test("register globalObject with custom name", () => {
    const localVue = createLocalVue();

    localVue.use(VueGtag, {
      globalObjectName: "foo",
      config: {
        id: 1,
      },
    });

    expect(window.foo).toBeDefined();
    expect(window.dataLayer).toBeDefined();
  });

  test("register dataLayer with custom name", () => {
    const localVue = createLocalVue();

    localVue.use(VueGtag, {
      globalDataLayerName: "bar",
      config: {
        id: 1,
      },
    });

    expect(window.gtag).toBeDefined();
    expect(window.bar).toBeDefined();
  });

  test("first hit is current date", () => {
    const localVue = createLocalVue();

    localVue.use(VueGtag, {
      config: {
        id: 1,
      },
    });

    expect(window.dataLayer).toMatchSnapshot();
  });

  test("optOut when disabled", async () => {
    const localVue = createLocalVue();

    localVue.use(VueGtag, {
      enabled: false,
      config: {
        id: 1,
      },
    });

    await flushPromises();

    expect(api.optOut).toHaveBeenCalled();
  });
});
