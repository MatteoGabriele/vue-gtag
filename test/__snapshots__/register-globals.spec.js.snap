// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerGlobals first hit is current date 1`] = `
Array [
  Arguments [
    "js",
    2021-04-20T10:00:00.000Z,
  ],
]
`;
